### PR TITLE
Add eth_getProof support for historical blocks

### PR DESCRIPTION
**Pull request by jyellick**
This PR starts with a few small commits of code cleanup.  Reviewed separately they should hopefully obviously be functionally no-ops.  I'm happy to strip these out and submit them separately if desired.

The final commit is to add support for older blocks as a parameter to eth_getProof.  In order to compute proofs, the function leverages the staged sync unwinding code to bring the hashed state table back to its historic state, as well as to build a list of trie nodes which need to be invalidated/re-computed in the trie computation.  Because these operations could be expensive for very old blocks, it limits the distance proofs are allowed from the head. It also adds some additional checks for correctness, as well as tests which verify the implementation.

This was discussed a bit on Discord in the db-format topic.

### CODE REVIEW

File: cmd/rpcdaemon/commands/eth_call.go [hunk 1]
```diff
@@ -312,59 +329,101 @@ func (api *APIImpl) GetProof(ctx context.Context, address libcommon.Address, sto
...
-		reader, err := rpchelper.CreateStateReader(ctx, tx, blockNrOrHash, 0, api.filters, api.stateCache, api.historyV3(tx), "")
-		if err != nil {
-			return nil, err
+	var loader *trie.FlatDBTrieLoader
+	if blockNr < latestBlock {
+		if latestBlock-blockNr > maxGetProofRewindBlockCount {
+			return nil, fmt.Errorf("requested block is too old, block must be within %d blocks of the head block number (currently %d)", maxGetProofRewindBlockCount, latestBlock)
 		}
-		a, err := reader.ReadAccountData(address)
+		tmpDir, err := os.MkdirTemp("", "eth_getHash")
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/eth_call.go:369]: i advise youse existing `tmpdir`: no reason to manually create/remove folders/files - memdb/etl - will automatically delete what need - if `defer rollback/close`.

File: cmd/rpcdaemon/commands/eth_call.go [hunk 2]
```diff
@@ -312,59 +329,101 @@ func (api *APIImpl) GetProof(ctx context.Context, address libcommon.Address, sto
...
-	} else {
-		addrHash, err := common.HashData(address[:])
-		if err != nil {
-			return nil, err
-		}
+	}
+
+	if latestBlock < blockNr {
+		// shouldn't happen, but check anyway
+		return nil, fmt.Errorf("block number is in the future")
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/eth_call.go:349]: let's add more info to the error: values of latestBlock, blockNr

File: cmd/rpcdaemon/commands/eth_call.go (hunk 1 above)
Comment (jyellick) [reply cmd/rpcdaemon/commands/eth_call.go]: I didn't like this part of the PR, but, wasn't really sure of the best way to do it and didn't see any of the other RPC daemon commands referencing a tempdir.

My best guess is that you're referring to the datadir temp dir?  I see there is a datadir on the CLI flags, so, I've pushed a commit on top that wires the `datadir.Dirs` through to the call.  Unfortunately this creates a bit of churn in the other tests, but, it's not too bad.

I'm not sure if that's what you were hoping for, so please let me know if you had something different in mind, happy to take a different approach.

File: cmd/rpcdaemon/commands/daemon.go [hunk 3]
```diff
@@ -18,7 +18,7 @@ func APIList(db kv.RoDB, borDb kv.RoDB, eth rpchelper.ApiBackend, txPool txpool.
 	blockReader services.FullBlockReader, agg *libstate.AggregatorV3, cfg httpcfg.HttpCfg, engine consensus.EngineReader,
 ) (list []rpc.API) {
 	base := NewBaseApi(filters, stateCache, blockReader, agg, cfg.WithDatadir, cfg.EvmCallTimeout, engine)
-	ethImpl := NewEthAPI(base, db, eth, txPool, mining, cfg.Gascap, cfg.ReturnDataLimit)
+	ethImpl := NewEthAPI(base, db, eth, txPool, mining, cfg.Gascap, cfg.ReturnDataLimit, cfg.Dirs)
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/daemon.go:21]: `dirs` are immutable, can add this field to BaseApi

File: cmd/rpcdaemon/commands/eth_call.go [hunk 4]
```diff
@@ -312,59 +327,94 @@ func (api *APIImpl) GetProof(ctx context.Context, address libcommon.Address, sto
 	}
 	defer tx.Rollback()
 
-	blockNr, _, _, err := rpchelper.GetBlockNumber(blockNrOrHash, tx, api.filters)
+	blockNr, blockHash, _, err := rpchelper.GetBlockNumber(blockNrOrHash, tx, api.filters)
 	if err != nil {
 		return nil, err
 	}
 
+	header := rawdb.ReadHeader(tx, blockHash, blockNr)
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/eth_call.go:335]: use `api._blockReader.HeaderByNumber`

Comment (jyellick) [reply cmd/rpcdaemon/commands/eth_call.go]: Done

File: cmd/rpcdaemon/commands/daemon.go (hunk 3 above)
Comment (jyellick) [reply cmd/rpcdaemon/commands/daemon.go]: Done, moved it to BaseApi.

**Review by AskAlexSharov — APPROVED** [review]

File: cmd/rpcdaemon/commands/call_traces_test.go [hunk 5]
```diff
@@ -166,7 +168,7 @@ func TestFilterNoAddresses(t *testing.T) {
 	}
 	agg := m.HistoryV3Components()
 	br := snapshotsync.NewBlockReaderWithSnapshots(m.BlockSnapshots, m.TransactionsV3)
-	api := NewTraceAPI(NewBaseApi(nil, kvcache.New(kvcache.DefaultCoherentConfig), br, agg, false, rpccfg.DefaultEvmCallTimeout, m.Engine), m.DB, &httpcfg.HttpCfg{})
+	api := NewTraceAPI(NewBaseApi(nil, kvcache.New(kvcache.DefaultCoherentConfig), br, agg, false, rpccfg.DefaultEvmCallTimeout, m.Engine, datadir.New(t.TempDir())), m.DB, &httpcfg.HttpCfg{})
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/call_traces_test.go:171]: Already have field: MockSentry.Dirs

**Pull request merged**